src/swupd/fullfiles.go

```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createRegularFullfile(input, name, output string) error {
+	// Ensure this is a regular file.
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a regular file but it is not", input)
+	}
+
+	// Create the uncompressed fullfile.
+	uncompressed, err := os.Create(output)
+	if err != nil {
+		return fmt.Errorf("couldn't create uncompressed fullfile: %s", err)
+	}
+	err = tarRegularFullfile(uncompressed, input, name, fi)
+	if err != nil {
+		return fmt.Errorf("couldn't archive the file %s: %s", input, err)
+	}
+	uncompressedSize, err := uncompressed.Seek(0, io.SeekEnd)
+	if err != nil {
+		return fmt.Errorf("couldn't find the size of uncompressed fullfile: %s", err)
+	}
+	err = uncompressed.Sync()
+	if err != nil {
+		return fmt.Errorf("couldn't sync uncompressed fullfile: %s", err)
+	}
+	defer uncompressed.Close()
+
+	if DebugFullfiles {
+		log.Printf("DEBUG: Creating fullfile %s for regular file %s (%d bytes)", name, input, fi.Size())
+		log.Printf("DEBUG: %s (%d bytes, uncompressed)", filepath.Base(output), uncompressedSize)
+	}
+
+	// Pick the best compression option (or no compression) for that specific fullfile.
+	best := ""
+	var bestSize int64 = uncompressedSize
+	for i, c := range fullfileCompressors {
+		candidate := fmt.Sprintf("%s.%d.%s", output, i, c.Name)
+		_, err := uncompressed.Seek(0, io.SeekStart)
+		if err != nil {
+			return fmt.Errorf("couldn't seek in fullfile %s: %s", input, err)
+		}
+		args := fullfileCompressArgs{
+			output:       candidate,
+			tarredReader: uncompressed,
+		}
+		size, err := c.Func(args)
+		if err != nil {
+			log.Printf("WARNING: couldn't compress %s using compressor %q: %s", input, c.Name, err)
+			os.RemoveAll(candidate)
+			continue
+		}
+		if size < bestSize {
+			if best != "" {
+				os.RemoveAll(best)
+			}
+			best = candidate
+			bestSize = size
+		} else {
+			os.RemoveAll(candidate)
+		}
+
+		if DebugFullfiles {
+			log.Printf("DEBUG: %s (%d bytes)", filepath.Base(candidate), size)
+		}
+	}
+
+	if best != "" {
+		// Failure during rename might indicate some further problems, so return error
+		// instead of ignoring it (since there is an uncompressed version).
+		return os.Rename(best, output)
+	}
+	return nil
+}
+
+func tarRegularFullfile(w io.Writer, input, name string, fi os.FileInfo) error {
+	tw := tar.NewWriter(w)
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return err
+	}
+	hdr.Typeflag = tar.TypeReg
+	hdr.Name = name
+	err = tw.WriteHeader(hdr)
+	if err != nil {
+		return err
+	}
+	in, err := os.Open(input)
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(tw, in)
+	if err != nil {
+		in.Close()
+		return err
+	}
+	in.Close()
+	return tw.Close()
```

Comment:
tarRegular ought to make an explicit commitment about the state of iostreams after it finishes. If you have errors then tw.Close is not called, so you leak a descriptor.

Reply:
It is up to the caller to Close the 'w', which in this case is backed by a file descriptor. 'tw' is not backed by a file descriptor so not really important here.

The leak comment is relevant but for the caller of this function. I'll fix that.